editor/import/3d/post_import_plugin_skeleton_rest_fixer.cpp

```diff
@@ -204,7 +204,7 @@ void PostImportPluginSkeletonRestFixer::internal_process(InternalImportCategory
 								} else {
 									for (int j = 0; j < key_len; j++) {
 										Basis sc = Basis().scaled(static_cast<Vector3>(anim->track_get_key_value(i, j)));
-										anim->track_set_key_value(i, j, (global_transform.basis * sc).get_scale());
+										anim->track_set_key_value(i, j, (global_transform.orthonormalized().basis * sc).get_scale());
```

Comment:
I assume this is not a valid solution for bones that are intentionally scaled as glTF, as there is no guarantee that this is normalized.

I think there is a guarantee that glTF bones exported from Blender will be scale normalized but it is not for all glTF.

So, if we do this, wouldn't we need to process the Apply Node Scale to set the Scale of all bones to 1?

Reply:
Why would it be invalid for scaled bones? global_transforms is the global scale adjustment propagated from nodes, and that is what is being orthonormalized. The point of Apply Node Transforms is to make these transforms identity. The code is not orthonormalizing the scale and so still apply any non-uniform local bone scales.

if your assertion is that this code is wrong, then line 146 (set_bone_pose_scale) would also be wrong. But I would need to see a test case

Reply:
I think line 136 is correct, but the code to apply scaling to each bone is lacking.

Also, I think the code for the scale track around line 200 is wrong: the position track is scaled, but the scale track should be multiplied by the inverse.

Reply:
For test cases, you should have tested only:

```
Skeleton: scale = (0.01, 0.01, 0.01)
L Hips: scale = (1.0, 1.0, 1.0)
  L Spine: scale = (1.0, 1.0, 1.0)
    L Chest: scale = (1.0, 1.0, 1.0)
```

So, the following should be tested:

```
Skeleton: scale = (1.0, 1.0, 1.0)
L Hips: scale = (0.1, 0.1, 0.1)
  L Spine: scale = (0.1, 0.1, 0.1)
    L Chest: scale = (1.0, 1.0, 1.0)
```
```
Skeleton: scale = (1.0, 1.0, 1.0)
L Hips: scale = (1.0, 1.0, 1.0)
  L Spine: scale = (0.1, 0.1, 0.1)
    L Chest: scale = (1.0, 1.0, 1.0)
```
```
Skeleton: scale = (1.0, 1.0, 1.0)
L Hips: scale = (1.0, 1.0, 1.0)
  L Spine: scale = (2.0, 1.0, -1.0)
    L Chest: scale = (1.0, 1.0, 1.0)
```

I guess the only restriction on the bone scale that glTF has is orthogonal. So I expect that scale(0, 0, 0) would not be acceptable, but something like scale(0.001, 0.001, 0.001) or scale(1, 2, 1) would be acceptable

Reply:
> Suppose we have a model with a neck rest scale of 0.001 for headless-able model like Dullahan, and an animation key value with a pose scale of 1.0 for the neck with a head for retargeting. If the rest scale of the neck is set to 1.0, then the animation key value should be 1000.0.

Ah, I missed that this is the PR to do that. If so, this change may be correct.

However, apart from this, I am wondering if the past remove immutable tracks are dead code due to the #90064 change, are they still being used for something other than Skeleton?

Reply:
Are you aware of the difference between orthonormal (uniform scale) and orthogonal (non uniform scale) methods on basis?

Reply:
https://github.com/godotengine/godot/blob/5708a3a02e00061e03366f2dabf8942df66fedca/core/math/basis.cpp#L80

Reply:
https://github.com/godotengine/godot/blob/5708a3a02e00061e03366f2dabf8942df66fedca/core/math/basis.cpp#L56

Reply:
Yes, I assume the glTF restriction is orthogonal with no isotropic and allows non-uniform scale, but please point out if I'm wrong. However, I am getting a bit off-topic.

I think we need to make a new proposal on how to handle individual bone scales.